File: tools/tracer/src/tracer/commands/info.py
"""`trace info` — complexity structure + architectural overview.

For a file: function list with cyclomatic complexity (lizard, cached via
file_facts), language, LOC, nearest doc, and the file's per-file facts
from the file/ cache.

For a directory: aggregated per-file stats sourced from the file/ cache
(no per-call re-extraction).
"""

from __future__ import annotations

import json
from pathlib import Path

import click
import lizard

from tracer import architecture, cache, digest, file_facts, passive_context, repo_files
from tracer.deps import require_dependencies
from tracer.enrich import nearest_doc
from tracer.repo_context import repo_context


def _file_info(path: Path) -> dict:
    """Per-file info; runs one lizard pass for per-function detail and
    derives the summary aggregates (function_count, ccn_total, ccn_max)
    from the same pass. file_facts is queried with cache_only=True so it
    doesn't trigger a second lizard run during cache miss."""
    repo_root = cache.repo_root_for(path)
    facts = file_facts.get(path, repo_root=repo_root, cache_only=True)
    try:
        parsed = lizard.analyze_file(str(path))
        functions = [
            {
                "name": f.name,
                "cyclomatic_complexity": f.cyclomatic_complexity,
                "nloc": f.nloc,
                "params": f.parameter_count,
                "start_line": f.start_line,
                "end_line": f.end_line,
            }
            for f in parsed.function_list
        ]
    except Exception:
        functions = []

    ccn_total = sum(f["cyclomatic_complexity"] for f in functions)
    ccn_max = max((f["cyclomatic_complexity"] for f in functions), default=0)
    function_count = len(functions)
    rank = "low" if ccn_total < 10 else "medium" if ccn_total < 30 else "high" if ccn_total < 80 else "critical"

    leading = digest.leading_comment(path)
    callers: list[dict] = []
    deps: list[dict] = []
    graph = architecture.load_cached(repo_root)
    if graph is not None:
        try:
            relative = str(path.resolve().relative_to(repo_root.resolve()))
        except ValueError:
            relative = str(path)
        callers = digest.top_callers(graph, relative, repo_root=repo_root)
        deps = digest.immediate_dependencies(graph, relative)

    return {
        "file": str(path),
        "language": facts.language if facts else None,
        "loc": facts.loc if facts else (functions[0]["nloc"] if functions else 0),
        "function_count": function_count,
        "cyclomatic_complexity_total": ccn_total,
        "cyclomatic_complexity_max": ccn_max,
        "rank": rank,
        "functions": functions,
        "nearest_doc": nearest_doc(path),
        "passive_context": passive_context.render(facts) if facts else None,
        "leading_comment": leading,
        "top_callers": callers,
        "dependencies": deps,
    }


def _dir_info(path: Path) -> dict:
    """Aggregate via `repo_files.tracked_files` inside a repo (respects
    .gitignore, never descends into ignored trees); fall back to
    `repo_files.walk_files` outside a repo. file_facts queried with
    cache_only=True — info never blocks on per-file extraction."""
    base = path.resolve()
    repo_root = cache.repo_root_for(base)
    tracked = repo_files.tracked_files(repo_root, base=base)

    files: list[dict] = []
    total_count = 0

    if tracked is not None:
        for rel in sorted(tracked):
            full = repo_root / rel
            try:
                under_base = str(full.resolve().relative_to(base))
            except ValueError:
                continue
            total_count += 1
            facts = file_facts.get(full, repo_root=repo_root, cache_only=True)
            if facts is None:
                continue
            files.append(
                {
                    "file": under_base,
                    "abs_path": str(full),
                    "loc": facts.loc,
                    "cyclomatic_complexity_total": facts.cyclomatic_complexity_total,
                    "function_count": facts.function_count,
                    "rank": facts.rank,
                    "passive_context": passive_context.render_compact(facts),
                }
            )
    else:
        for full in sorted(repo_files.walk_files(base)):
            total_count += 1
            facts = file_facts.get(full, repo_root=repo_root, cache_only=True)
            if facts is None:
                continue
            files.append(
                {
                    "file": str(full.relative_to(base)),
                    "abs_path": str(full),
                    "loc": facts.loc,
                    "cyclomatic_complexity_total": facts.cyclomatic_complexity_total,
                    "function_count": facts.function_count,
                    "rank": facts.rank,
                    "passive_context": passive_context.render_compact(facts),
                }
            )

    return {
        "directory": str(path),
        "file_count": total_count,
        "cyclomatic_complexity_total": sum(f["cyclomatic_complexity_total"] for f in files),
        "loc_total": sum(f["loc"] for f in files),
        "files": files,
        "nearest_doc": nearest_doc(path),
    }


@click.command()
@click.argument("path", type=click.Path(exists=True))
@click.option("--json", "as_json", is_flag=True)
@click.option("--brief", is_flag=True, help="Show only top-3 most-complex functions instead of all.")
def command(path: str, as_json: bool, brief: bool) -> None:
    """Complexity structure + architectural overview of a file or directory.

    Default output is full: purpose, top callers, dependencies, lifecycle,
    and the complete function table. `--brief` truncates the function table
    to the 3 hottest functions.
    """
    full = not brief
    require_dependencies()
    p = Path(path).resolve()
    info = _file_info(p) if p.is_file() else _dir_info(p)
    info["repo_context"] = repo_context(str(p))

    if as_json:
        click.echo(json.dumps(info, indent=2))
        return

    if p.is_file():
        click.echo(f"File: {info['file']}")
        if info.get("passive_context"):
            click.echo(info["passive_context"])
        click.echo(f"Language: {info.get('language')}")
        click.echo(
            f"LOC: {info['loc']}  Functions: {info['function_count']}  "
            f"CCN total: {info['cyclomatic_complexity_total']}  "
            f"CCN max: {info['cyclomatic_complexity_max']}  Rank: {info['rank']}"
        )
        click.echo(f"Nearest doc: {info.get('nearest_doc') or '(none)'}")
        if info.get("leading_comment"):
            click.echo("")
            click.echo("Purpose (from leading comment):")
            for line in info["leading_comment"].splitlines():
                click.echo(f"  {line}")
        if info.get("top_callers"):
            click.echo("")
            click.echo("Top callers (modules depending on this file):")
            for caller in info["top_callers"]:
                if caller.get("source_line"):
                    where = f" — {caller['source_file']}:{caller['source_line']}"
                elif caller.get("source_file"):
                    where = f" — {caller['source_file']}"
                else:
                    where = ""
                summary = caller.get("summary")
                summary_suffix = f"  {summary}" if summary else ""
                click.echo(f"  {caller['label']} ({caller['kind']}){where}{summary_suffix}")
        if info.get("dependencies"):
            click.echo("")
            click.echo("Immediate dependencies (modules this file imports):")
            for dep in info["dependencies"]:
                click.echo(f"  {dep['module']}  [{dep['confidence']}]")
        click.echo("")
        # Brief default: top-3 hottest functions; --full shows the whole table.
        top_funcs = sorted(info["functions"], key=lambda f: -f["cyclomatic_complexity"])
        shown = top_funcs if full else top_funcs[:3]
        click.echo(f"Functions ({'all' if full else 'top 3 by complexity'} of {len(info['functions'])}):")
        for f in shown:
            click.echo(
                f"  {f['cyclomatic_complexity']:>3}  {f['nloc']:>4} loc  "
                f"{f['name']}({f['params']} params)  L{f['start_line']}-{f['end_line']}"
            )
        if not full and len(info["functions"]) > 3:
            click.echo(f"  … {len(info['functions']) - 3} more (use --full to see all)")
    else:
        click.echo(f"Directory: {info['directory']}")
        click.echo(
            f"Files: {info['file_count']}  LOC: {info['loc_total']}  "
            f"CCN total: {info['cyclomatic_complexity_total']}"
        )
        click.echo(f"Nearest doc: {info.get('nearest_doc') or '(none)'}")
        click.echo("")
        click.echo("Files (top 20 by complexity, with file digest):")
        for f in sorted(info["files"], key=lambda x: -x["cyclomatic_complexity_total"])[:20]:
            click.echo(
                f"  {f['cyclomatic_complexity_total']:>5}  {f['loc']:>5} loc  "
                f"{f['function_count']:>3} fn  [{f['rank']:<8}]  "
                f"{f['file']}  ({f.get('passive_context','')})"
            )
            abs_path = f.get("abs_path")
            if abs_path:
                purpose = digest.leading_comment(Path(abs_path))
                if purpose:
                    first_line = purpose.splitlines()[0].strip()
                    if first_line:
                        click.echo(f"        Purpose: {first_line}")
                try:
                    parsed = lizard.analyze_file(abs_path)
                    fns = sorted(parsed.function_list, key=lambda x: -x.cyclomatic_complexity)[:3]
                    if fns:
                        sig_line = "  ".join(
                            f"{fn.name}() L{fn.start_line}" for fn in fns
                        )
                        click.echo(f"        Top fns: {sig_line}")
                except Exception:
                    pass

    ctx = info["repo_context"]
    click.echo("")
    click.echo(
        f"repo_context: complexity_p95={ctx['complexity_p95']} "
        f"median={ctx['median_file_ccn']} files={ctx['total_files']}"
    )

File: packages/claude/bin/tracer-dist/crate/src/commands/info.rs
//! `trace info` — complexity structure + architectural overview.
//! The per-function table is AST-derived; `nloc` is the line span.
//! Emits a function complexity profile plus architecture context.

use crate::{architecture, cache, ccn, file_facts, passive_context, repo_context};
use anyhow::Result;
use serde_json::{json, Value};
use std::path::Path;

fn rank(ccn_total: i64) -> &'static str {
    if ccn_total < 10 {
        "low"
    } else if ccn_total < 30 {
        "medium"
    } else if ccn_total < 80 {
        "high"
    } else {
        "critical"
    }
}

fn leading_comment(path: &Path) -> Option<String> {
    crate::digest::leading_comment(path, 25)
}

fn file_info(path: &Path) -> Value {
    let repo_root = cache::worktree_root_for(path).unwrap_or_else(|| cache::display_root(path));
    let facts = file_facts::get(path, &repo_root, None);

    let source = std::fs::read(path).unwrap_or_default();
    let functions = ccn::analyze(&source, &path.to_string_lossy()).unwrap_or_default();
    let fn_json: Vec<Value> = functions
        .iter()
        .map(|f| {
            json!({
                "name": f.name,
                "cyclomatic_complexity": f.cyclomatic_complexity,
                "nloc": f.nloc,
                "start_line": f.start_line,
                "end_line": f.start_line + f.nloc - 1,
            })
        })
        .collect();

    let ccn_total: i64 = functions.iter().map(|f| f.cyclomatic_complexity).sum();
    let ccn_max: i64 = functions
        .iter()
        .map(|f| f.cyclomatic_complexity)
        .max()
        .unwrap_or(0);
    let function_count = functions.len() as i64;

    let leading = leading_comment(path);
    let mut callers: Vec<Value> = vec![];
    let mut deps: Vec<Value> = vec![];
    if let Some(graph) = architecture::load_cached(&repo_root) {
        let relative = cache::relative_to_root(path, &repo_root);
        callers = crate::digest::top_callers(&graph, &relative, Some(&repo_root), 10);
        deps = crate::digest::immediate_dependencies(&graph, &relative, 15);
    }

    let loc = facts
        .as_ref()
        .map(|f| f.loc)
        .unwrap_or_else(|| functions.first().map(|f| f.nloc).unwrap_or(0));

    json!({
        "file": path.to_string_lossy(),
        "language": facts.as_ref().and_then(|f| f.language.clone()),
        "loc": loc,
        "function_count": function_count,
        "cyclomatic_complexity_total": ccn_total,
        "cyclomatic_complexity_max": ccn_max,
        "rank": rank(ccn_total),
        "functions": fn_json,
        "nearest_doc": crate::digest::nearest_doc(path),
        "passive_context": facts.as_ref().map(|f| passive_context::render(f, None)),
        "leading_comment": leading,
        "top_callers": callers,
        "dependencies": deps,
    })
}

fn dir_info(path: &Path) -> Value {
    let base = cache::absolutize(path);
    let repo_root = cache::worktree_root_for(&base).unwrap_or_else(|| cache::display_root(&base));
    let tracked = crate::repo_files::tracked_files(&repo_root, Some(&base));

    let mut files: Vec<Value> = vec![];
    let mut total_count = 0i64;

    let push_facts = |under_base: String,
                       full: &Path,
                       f: &file_facts::FileFacts,
                       files: &mut Vec<Value>| {
        files.push(json!({
            "file": under_base,
            "abs_path": full.to_string_lossy(),
            "loc": f.loc,
            "cyclomatic_complexity_total": f.cyclomatic_complexity_total,
            "function_count": f.function_count,
            "rank": f.rank,
            "passive_context": passive_context::render_compact(f),
        }));
    };

    match tracked {
        Some(rels) => {
            let mut rels = rels;
            rels.sort();
            // SINGLE batch (was O(N²) per-file get() — the dotfiles
            // info-dir cold regression #4). Real extraction preserved.
            let fulls: Vec<std::path::PathBuf> =
                rels.iter().map(|r| repo_root.join(r)).collect();
            let facts_map = file_facts::get_batch(&fulls, &repo_root);
            for rel in &rels {
                let full = repo_root.join(rel);
                let under = match full
                    .canonicalize()
                    .ok()
                    .and_then(|c| c.strip_prefix(&base).ok().map(|p| p.to_path_buf()))
                {
                    Some(p) => p.to_string_lossy().to_string(),
                    None => continue,
                };
                total_count += 1;
                if let Some(f) = facts_map.get(rel) {
                    push_facts(under, &full, f, &mut files);
                }
            }
        }
        None => {
            let mut walked = crate::repo_files::walk_files(&base);
            walked.sort();
            let facts_map = file_facts::get_batch(&walked, &repo_root);
            for full in &walked {
                total_count += 1;
                let under = full
                    .strip_prefix(&base)
                    .map(|p| p.to_string_lossy().to_string())
                    .unwrap_or_else(|_| full.to_string_lossy().to_string());
                let rel = cache::relative_to_root(full, &repo_root);
                if let Some(f) = facts_map.get(&rel) {
                    push_facts(under, full, f, &mut files);
                }
            }
        }
    }

    let ccn_total: i64 = files
        .iter()
        .map(|f| f["cyclomatic_complexity_total"].as_i64().unwrap_or(0))
        .sum();
    let loc_total: i64 = files.iter().map(|f| f["loc"].as_i64().unwrap_or(0)).sum();

    json!({
        "directory": path.to_string_lossy(),
        "file_count": total_count,
        "cyclomatic_complexity_total": ccn_total,
        "loc_total": loc_total,
        "files": files,
        "nearest_doc": crate::digest::nearest_doc(path),
    })
}

pub fn run(path: &Path, as_json: bool, brief: bool) -> Result<Value> {
    crate::pathval::require_exists(path, "PATH");
    let p = cache::absolutize(path);
    let mut info = if p.is_file() {
        file_info(&p)
    } else {
        dir_info(&p)
    };
    info["repo_context"] = repo_context::repo_context(&p);

    if !as_json {
        if p.is_file() {
            emit_file_human(&info, !brief);
        } else {
            emit_dir_human(&info);
        }
        let ctx = &info["repo_context"];
        println!();
        println!(
            "repo_context: complexity_p95={} median={} files={}",
            ctx["complexity_p95"].as_i64().unwrap_or(0),
            ctx["median_file_ccn"].as_i64().unwrap_or(0),
            ctx["total_files"].as_i64().unwrap_or(0),
        );
    }
    Ok(info)
}

fn emit_file_human(info: &Value, full: bool) {
    println!("File: {}", info["file"].as_str().unwrap_or(""));
    if let Some(pc) = info["passive_context"].as_str() {
        println!("{pc}");
    }
    println!(
        "Language: {}",
        info["language"].as_str().unwrap_or("None")
    );
    println!(
        "LOC: {}  Functions: {}  CCN total: {}  CCN max: {}  Rank: {}",
        info["loc"].as_i64().unwrap_or(0),
        info["function_count"].as_i64().unwrap_or(0),
        info["cyclomatic_complexity_total"].as_i64().unwrap_or(0),
        info["cyclomatic_complexity_max"].as_i64().unwrap_or(0),
        info["rank"].as_str().unwrap_or(""),
    );
    println!(
        "Nearest doc: {}",
        info["nearest_doc"].as_str().unwrap_or("(none)")
    );
    if let Some(lc) = info["leading_comment"].as_str() {
        println!();
        println!("Purpose (from leading comment):");
        for line in lc.lines() {
            println!("  {line}");
        }
    }
    if let Some(callers) = info["top_callers"].as_array() {
        if !callers.is_empty() {
            println!();
            println!("Top callers (modules depending on this file):");
            for c in callers {
                let label = c["label"].as_str().unwrap_or("");
                let kind = c["kind"].as_str().unwrap_or("");
                let where_ = match (c["source_line"].as_i64(), c["source_file"].as_str())
                {
                    (Some(l), Some(f)) => format!(" — {f}:{l}"),
                    (None, Some(f)) => format!(" — {f}"),
                    _ => String::new(),
                };
                let summary = c["summary"]
                    .as_str()
                    .map(|s| format!("  {s}"))
                    .unwrap_or_default();
                println!("  {label} ({kind}){where_}{summary}");
            }
        }
    }
    if let Some(deps) = info["dependencies"].as_array() {
        if !deps.is_empty() {
            println!();
            println!("Immediate dependencies (modules this file imports):");
            for d in deps {
                println!(
                    "  {}  [{}]",
                    d["module"].as_str().unwrap_or(""),
                    d["confidence"].as_str().unwrap_or("")
                );
            }
        }
    }
    println!();
    let mut funcs: Vec<&Value> =
        info["functions"].as_array().map(|a| a.iter().collect()).unwrap_or_default();
    funcs.sort_by(|a, b| {
        b["cyclomatic_complexity"]
            .as_i64()
            .cmp(&a["cyclomatic_complexity"].as_i64())
    });
    let total = funcs.len();
    let shown: Vec<&Value> = if full {
        funcs.clone()
    } else {
        funcs.iter().take(3).cloned().collect()
    };
    println!(
        "Functions ({} of {}):",
        if full { "all" } else { "top 3 by complexity" },
        total
    );
    for f in shown {
        println!(
            "  {:>3}  {:>4} loc  {}  L{}-{}",
            f["cyclomatic_complexity"].as_i64().unwrap_or(0),
            f["nloc"].as_i64().unwrap_or(0),
            f["name"].as_str().unwrap_or(""),
            f["start_line"].as_i64().unwrap_or(0),
            f["end_line"].as_i64().unwrap_or(0),
        );
    }
    if !full && total > 3 {
        println!("  … {} more (use --full to see all)", total - 3);
    }
}

fn emit_dir_human(info: &Value) {
    println!("Directory: {}", info["directory"].as_str().unwrap_or(""));
    println!(
        "Files: {}  LOC: {}  CCN total: {}",
        info["file_count"].as_i64().unwrap_or(0),
        info["loc_total"].as_i64().unwrap_or(0),
        info["cyclomatic_complexity_total"].as_i64().unwrap_or(0),
    );
    println!(
        "Nearest doc: {}",
        info["nearest_doc"].as_str().unwrap_or("(none)")
    );
    println!();
    println!("Files (top 20 by complexity, with file digest):");
    let mut files: Vec<&Value> =
        info["files"].as_array().map(|a| a.iter().collect()).unwrap_or_default();
    files.sort_by(|a, b| {
        b["cyclomatic_complexity_total"]
            .as_i64()
            .cmp(&a["cyclomatic_complexity_total"].as_i64())
    });
    for f in files.iter().take(20) {
        println!(
            "  {:>5}  {:>5} loc  {:>3} fn  [{:<8}]  {}  ({})",
            f["cyclomatic_complexity_total"].as_i64().unwrap_or(0),
            f["loc"].as_i64().unwrap_or(0),
            f["function_count"].as_i64().unwrap_or(0),
            f["rank"].as_str().unwrap_or(""),
            f["file"].as_str().unwrap_or(""),
            f["passive_context"].as_str().unwrap_or(""),
        );
        // Per-file digest block: Purpose + Top fns (the 3 hottest
        // functions with their start lines, from the AST backend).
        if let Some(abs) = f["abs_path"].as_str() {
            let p = Path::new(abs);
            if let Some(purpose) = crate::digest::leading_comment(p, 25) {
                if let Some(first) = purpose.lines().next() {
                    let first = first.trim();
                    if !first.is_empty() {
                        println!("        Purpose: {first}");
                    }
                }
            }
            let src = std::fs::read(p).unwrap_or_default();
            if let Some(mut fns) = ccn::analyze(&src, abs) {
                fns.sort_by(|a, b| {
                    b.cyclomatic_complexity.cmp(&a.cyclomatic_complexity)
                });
                let top: Vec<String> = fns
                    .iter()
                    .take(3)
                    .map(|fn_| format!("{}() L{}", fn_.name, fn_.start_line))
                    .collect();
                if !top.is_empty() {
                    println!("        Top fns: {}", top.join("  "));
                }
            }
        }
    }
}
